apisix/cli/etcd.lua

```diff
@@ -0,0 +1,205 @@
+    end
+
+    if yaml_conf.apisix.config_center ~= "etcd" then
+        return true
+    end
+
+    if not yaml_conf.etcd then
+        util.die("failed to read `etcd` field from yaml file when init etcd")
+    end
+
+    local etcd_conf = yaml_conf.etcd
+
+    local timeout = etcd_conf.timeout or 3
+    local uri
+
+    -- convert old single etcd config to multiple etcd config
+    if type(yaml_conf.etcd.host) == "string" then
+        yaml_conf.etcd.host = {yaml_conf.etcd.host}
+    end
+
+    local host_count = #(yaml_conf.etcd.host)
+    local scheme
+    for i = 1, host_count do
+        local host = yaml_conf.etcd.host[i]
+        local fields = util.split(host, "://")
+        if not fields then
+            util.die("malformed etcd endpoint: ", host, "\n")
+        end
+
+        if not scheme then
+            scheme = fields[1]
+        elseif scheme ~= fields[1] then
+            print([[WARNING: mixed protocols among etcd endpoints]])
+        end
+    end
+
+    -- check the etcd cluster version
+    for index, host in ipairs(yaml_conf.etcd.host) do
+        uri = host .. "/version"
+        local cmd = str_format("curl -s -m %d %s", timeout * 2, uri)
+        local res = util.execute_cmd(cmd)
+        local errmsg = str_format("got malformed version message: \"%s\" from etcd\n",
+                                  res)
+
+        local body, _, err = dkjson.decode(res)
+        if err then
+            util.die(errmsg)
+        end
+
+        local cluster_version = body["etcdcluster"]
+        if not cluster_version then
+            util.die(errmsg)
+        end
+
+        if compare_semantic_version(cluster_version, env.min_etcd_version) then
+            util.die("etcd cluster version ", cluster_version,
+                     " is less than the required version ",
+                     env.min_etcd_version,
+                     ", please upgrade your etcd cluster\n")
+        end
+    end
+
+    local etcd_ok = false
+    for index, host in ipairs(yaml_conf.etcd.host) do
+        local is_success = true
+
+        for _, dir_name in ipairs({"/routes", "/upstreams", "/services",
+                                   "/plugins", "/consumers", "/node_status",
+                                   "/ssl", "/global_rules", "/stream_routes",
+                                   "/proto", "/plugin_metadata"}) do
+
+            local key =  (etcd_conf.prefix or "") .. dir_name .. "/"
+
+            local uri = host .. "/v3/kv/put"
+            local post_json = '{"value":"' .. base64_encode("init_dir")
+                              ..'", "key":"' .. base64_encode(key) .. '"}'
+            local cmd = "curl " .. uri .. " -X POST -d '" .. post_json
+                        .. "' --connect-timeout " .. timeout
+                        .. " --max-time " .. timeout * 2 .. " --retry 1 2>&1"
+
+            local res = util.execute_cmd(cmd)
+            if res:find("OK", 1, true) then
```

Comment:
The error check is confusing to me... Why "OK" means "is_success" is false?

Reply:
Me too... @Yiyiyimu

Reply:
![image](https://user-images.githubusercontent.com/10428333/99481070-4ab1d880-2994-11eb-85d6-ca536494cc34.png)

The output should like this.

Reply:
I can't reproduce why I did so. The original code before upgrading to etcd v3 for implementing this is 
https://github.com/apache/apisix/blob/c54aec8f6cec2bd31c7c8717bf5220327584be03/bin/apisix#L899-L907

Maybe that could be of help

Reply:
Let me just adjust it.

Reply:
Fixed